libtransmission/torrent.cc

```diff
@@ -2352,20 +2352,31 @@ void tr_torrentFileCompleted(tr_torrent* tor, tr_file_index_t i)
     }
 }
 
-void tr_torrentPieceCompleted(tr_torrent* tor, tr_piece_index_t piece_index)
+void onPieceCompleted(tr_torrent* tor, tr_piece_index_t piece)
 {
-    tr_peerMgrPieceCompleted(tor, piece_index);
+    tr_peerMgrPieceCompleted(tor, piece);
 
     // if this piece completes any file, invoke the fileCompleted func for it
-    auto const span = tor->fpm_.fileSpan(piece_index);
+    auto const span = tor->fpm_.fileSpan(piece);
     for (auto file = span.begin; file < span.end; ++file)
     {
         if (tor->completion.hasBlocks(tr_torGetFileBlockSpan(tor, file)))
         {
-            tr_torrentFileCompleted(tor, file);
+            onFileCompleted(tor, file);
         }
     }
 }
+
+void onPieceFailed(tr_torrent* tor, tr_piece_index_t piece)
+{
+    tr_logAddDebugTor(tor, fmt::format("Piece {}, which was just downloaded, failed its checksum test", piece));
+
+    auto const n = tor->pieceSize(piece);
+    tor->corruptCur += n;
+    tor->downloadedCur -= std::min(tor->downloadedCur, uint64_t{ n });
+    tr_peerMgrGotBadPiece(tor, piece);
+    tor->setHasPiece(piece, false);
```

Comment:
So essentially, that one line, `tor->setHasPiece(piece, false);` is the expected fix, right? (the rest being refactoring)

But considering that `setHasPiece` is only called in a single place in the whole codebase, which is `tr_verify_worker::verifyTorrent`, shouldn't the default value be false already, or correctly evaluated during the verifyTorrent process?

Reply:
> But considering that setHasPiece is otherwise only called in a single place in the whole codebase, which is tr_verify_worker::verifyTorrent, shouldn't the default value be false already, or correctly evaluated during the verifyTorrent process?

No. The completion bitfield counts blocks, not pieecs. `setHasPiece()` invalidates the range of blocks that are part of that piece. Blocks are added when they come in via `addBlock()`.

Reply:
OK. I'll leave it to others to verify the effectiveness of the fix, but otherwise the code change is safe.